driver/src/orderbook/streamed/state.rs

```diff
@@ -0,0 +1,531 @@
+            Event::SolutionSubmission(_event) => unimplemented!(),
+        }
+    }
+
+    fn deposit(&mut self, event: &Deposit, block_batch_id: BatchId) -> Result<(), Error> {
+        let balance = self.balances.entry((event.user, event.token)).or_default();
+        balance.deposit(
+            Flux {
+                amount: event.amount,
+                batch_id: event.batch_id,
+            },
+            block_batch_id,
+        );
+        Ok(())
+    }
+
+    fn withdraw_request(&mut self, event: &WithdrawRequest) -> Result<(), Error> {
+        let balance = self.balances.entry((event.user, event.token)).or_default();
+        balance.pending_withdraw = Some(Flux {
+            amount: event.amount,
+            batch_id: event.batch_id,
+        });
+        Ok(())
+    }
+
+    fn withdraw(&mut self, event: &Withdraw, block_batch_id: BatchId) -> Result<(), Error> {
+        let balance = self.balances.entry((event.user, event.token)).or_default();
+        balance.withdraw(event.amount, block_batch_id)
+    }
+
+    fn token_listing(&mut self, event: &TokenListing) -> Result<(), Error> {
+        self.tokens.0.push((event.id, event.token));
+        Ok(())
+    }
+
+    fn order_placement(&mut self, event: &OrderPlacement) -> Result<(), Error> {
+        let order = Order {
+            buy_token: event.buy_token,
+            sell_token: event.sell_token,
+            valid_from: event.valid_from,
+            valid_until: event.valid_until,
+            price_numerator: event.price_numerator,
+            price_denominator: event.price_denominator,
+            used_amount: 0,
+        };
+        match self.orders.entry((event.owner, event.index)) {
+            Entry::Vacant(entry) => entry.insert(order),
+            Entry::Occupied(_) => return Err(Error::OrderAlreadyExists),
+        };
+        Ok(())
+    }
+
+    fn order_cancellation(
+        &mut self,
+        event: &OrderCancellation,
+        block_batch_id: BatchId,
+    ) -> Result<(), Error> {
+        let order = self
+            .orders
+            .get_mut(&(event.owner, event.id))
+            .ok_or(Error::UnknownOrder)?;
+        order.valid_until = block_batch_id - 1;
+        Ok(())
+    }
+
+    fn order_deletion(&mut self, event: &OrderDeletion) -> Result<(), Error> {
+        // Orders are allowed to be deleted multiple times.
+        self.orders.remove(&(event.owner, event.id));
+        Ok(())
+    }
+}
+
+#[derive(Clone, Copy, Debug, Error)]
+pub enum Error {
+    #[error("unknown token")]
+    UnknownToken,
+    #[error("unknown order")]
+    UnknownOrder,
+    #[error("order already exists")]
+    OrderAlreadyExists,
+    #[error("math underflow")]
+    MathUnderflow,
+}
+
+/// Bidirectional map between token id and token address.
+///
+/// Std does not have this type so we use a vector. Alternatively we could find a crate.
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+struct Tokens(Vec<(TokenId, TokenAddress)>);
```

Comment:
This could be done with `(HashMap<TokenId, TokenAddress>, HashMap<TokenAddress, TokenId>)`, but I'm not sure its worth it (not for the size of the token list we currently have at least).

Reply:
Definitely not worth it.